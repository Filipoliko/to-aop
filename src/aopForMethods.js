import createCallTrap from './trap/createCallTrap';
import { AOP_HOOKS } from './symbol';

export default function aopForMethods(target, pattern) {
  let original = {};
  let prototype = target.prototype;
  while (prototype) {
    Object.entries(Object.getOwnPropertyDescriptors(prototype)).forEach(
      function([property]) {
        try {
          if (property in original) {
            return;
          }

          let aopHooks = prototype[property][AOP_HOOKS];
          if (aopHooks) {
            const { object } = aopHooks[aopHooks.length - 1];
            aopHooks.push({
              target,
              object,
              property,
              pattern
            });
            return;
          }

          if (typeof prototype[property] === 'function') {
            original[property] = prototype[property];
            prototype[property] = createCallTrap({
              target,
              object: original,
              property,
              pattern
            });
          }
        } catch (_) {
          console.error(_);
        } // eslint-disable-line no-empty
      }
    );

    prototype = Reflect.getPrototypeOf(prototype);
  }
}
