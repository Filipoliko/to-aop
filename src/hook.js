export function createHook(name, regular, callback) {
  return {
    [name]: function hook(meta) {
      return hookFor(meta, regular, callback);
    }
  };
}

export function hookFor(meta, filterRule, callback) {
  if (typeof filterRule === 'string') {
    if (meta.property.includes(filterRule)) {
      return callback(meta);
    }

    return null;
  }
  if (filterRule instanceof RegExp) {
    if (filterRule.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof filterRule === 'function') {
    if (filterRule(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(filterRule)) {
    return filterRule.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof filterRule}. Method accept string, regexp, function and array.`
  );
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter'
});
