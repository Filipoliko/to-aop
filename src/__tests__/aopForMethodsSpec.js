import aopForMethods from '../aopForMethods';
import { hookName, createHook } from '../hook';
import createClasses from './createClasses';
import createPattern from './createPattern';

describe('for class', () => {
  let pattern = null;
  let pattern2 = null;
  let arroundPattern = null;

  describe('class method', () => {
    let beforeMethod = null;
    let afterMethod = null;
    let aroundMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();
      aroundMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
      pattern2 = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
      arroundPattern = createHook(hookName.aroundMethod, /.*/, aroundMethod);
    });

    it('should call pattern.aroundMethod', () => {
      let { A } = createClasses();

      aopForMethods(A, arroundPattern);
      new A('method').method({}, 1);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      let { A } = createClasses();

      aopForMethods(A, pattern);
      new A('method').method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect', () => {
      let { A, C } = createClasses();

      aopForMethods(A, pattern);
      aopForMethods(C, pattern);

      new C('method').method2({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.arroundMethod for extended class with multiple aspect', () => {
      let { A, C } = createClasses();

      aopForMethods(A, arroundPattern);
      aopForMethods(C, arroundPattern);
      new C('method').method2({}, 1);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended class', () => {
      let { B } = createClasses();

      aopForMethods(B, pattern);
      new B('method').method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod after class is instanced', () => {
      let { B } = createClasses();

      const b = new B('method');
      aopForMethods(B, pattern);
      b.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor', () => {
      let { B, C, A } = createClasses();

      const b = new B('method');
      aopForMethods(A, pattern);
      aopForMethods(A, pattern2);
      aopForMethods(C, pattern);
      aopForMethods(C, pattern2);
      aopForMethods(B, pattern);
      aopForMethods(B, pattern2);

      expect(b.method({}, 1)).toEqual('B method');

      expect(beforeMethod.mock.calls.length).toEqual(6);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(6);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for super.superMethod', () => {
      let { C } = createClasses();
      pattern = createPattern(
        {
          [hookName.beforeMethod]: 'superMethod',
          [hookName.afterMethod]: 'superMethod',
        },
        {
          [hookName.beforeMethod]: beforeMethod,
          [hookName.afterMethod]: afterMethod,
        }
      );

      aopForMethods(C, pattern);
      const c = new C('super method');

      expect(c.superMethod()).toEqual('super method');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.aroundMethod for super.superMethod', () => {
      let { C } = createClasses();

      aopForMethods(C, arroundPattern);
      const c = new C('super method');

      expect(c.superMethod()).toEqual(undefined);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('es5 problems', () => {
    let beforeMethod = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
    });

    it('should be same instance', () => {
      let { B, A } = createClasses();

      const b = new B('method');
      aopForMethods(B, pattern);
      b.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(b instanceof A).toBeTruthy();
      expect(b instanceof B).toBeTruthy();
    });
  });
});
