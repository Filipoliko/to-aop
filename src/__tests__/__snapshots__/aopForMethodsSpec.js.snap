// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aopForMethods class method should call pattern.aroundMethod 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "payload": undefined,
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.aroundMethod for super.superMethod 1`] = `
Array [
  Object {
    "args": Array [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "payload": undefined,
    "property": "superMethod",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.aroundMethod three times and modify payload 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "payload": undefined,
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.arroundMethod for extended class with multiple aspect 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "payload": undefined,
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod after class is instanced 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod after class is instanced 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "payload": Map {},
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": Object {},
    "object": Object {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for super.superMethod 1`] = `
Array [
  Object {
    "args": Array [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": Object {},
    "object": Object {
      "__defineGetter__": [Function],
      "__defineSetter__": [Function],
      "__lookupGetter__": [Function],
      "__lookupSetter__": [Function],
      "_privateMethod": [Function],
      "constructor": [Function],
      "hasOwnProperty": [Function],
      "isPrototypeOf": [Function],
      "method": [Function],
      "method2": [Function],
      "propertyIsEnumerable": [Function],
      "superMethod": [Function],
      "toLocaleString": [Function],
      "toString": [Function],
      "valueOf": [Function],
    },
    "property": "superMethod",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for super.superMethod 2`] = `
Array [
  Object {
    "args": Array [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": Object {},
    "object": Object {
      "__defineGetter__": [Function],
      "__defineSetter__": [Function],
      "__lookupGetter__": [Function],
      "__lookupSetter__": [Function],
      "_privateMethod": [Function],
      "constructor": [Function],
      "hasOwnProperty": [Function],
      "isPrototypeOf": [Function],
      "method": [Function],
      "method2": [Function],
      "propertyIsEnumerable": [Function],
      "superMethod": [Function],
      "toLocaleString": [Function],
      "toString": [Function],
      "valueOf": [Function],
    },
    "payload": "super method",
    "property": "superMethod",
    "target": [Function],
  },
]
`;
